File: Dongle Interface/DongleSerialInterfaceParser.cs
using System.Text;

namespace QSenseExamples
{
    public static class DongleSerialInterfaceParser
    {
        private const Int32 MAX_NODES = 13;
        private const string DEVICE_NAME = "QSense\0";
        private const char PREFIX = '$';
        private const char POSTFIX = '\n';
        private const char OPCODE_TRANSMIT = 'T';
        private const char OPCODE_RECEIVE = 'R';
        private const char OPCODE_CONNECT = 'C';
        private const char OPCODE_CONNECT_WHITELIST = 'W';
        private const char OPCODE_DISCONNECT = 'D';
        private const char OPCODE_STOPSCAN = 'I';
        private const char OPCODE_STATUS = 'S';

        public static string ByteArrayToString(byte[] data)
        {
            string hexString = "";
            foreach (byte b in data) hexString += b.ToString("X2");
            return hexString;
        }

        public static string CreateTransmitPacket(byte[] coreInterfacePacket, int handle)
        {
            string hexString = ByteArrayToString(coreInterfacePacket);
            return "" + PREFIX + OPCODE_TRANSMIT + handle.ToString("X2") + hexString + POSTFIX;
        }

        public static string CreateConnectPacket()
        {
            string packet = "" + PREFIX + OPCODE_CONNECT;
            packet += MAX_NODES.ToString("X2");
            byte[] buffer = ASCIIEncoding.ASCII.GetBytes(DEVICE_NAME);
            foreach (byte b in buffer) packet += b.ToString("X2");
            return packet;
        }

        public static string CreateConnectWhitelistPacket(string[] serialNumbers)
        {
            string packet = "" + PREFIX + OPCODE_CONNECT_WHITELIST;
            packet += (serialNumbers.Length).ToString("X2");
            UInt64[] macAddresses = serialNumbers.Select(sn => Convert.ToUInt64(sn.Split("-")[0], 16)).ToArray();
            foreach (UInt64 macAddress in macAddresses)
            {
                byte[] buffer = new byte[6];
                Array.Copy(BitConverter.GetBytes(macAddress), 0, buffer, 0, 6);
                foreach (byte b in buffer) packet += b.ToString("X2");
            }
            packet += POSTFIX;
            return packet;
        }

        public static string CreateDisconnectPacket()
        {
            return "" + PREFIX + OPCODE_DISCONNECT + POSTFIX;
        }

        public static string CreateStopScanPacket()
        {
            return "" + PREFIX + OPCODE_STOPSCAN + POSTFIX;
        }

        public static string CreateStatusRequest()
        {
            return "" + PREFIX + OPCODE_STATUS + POSTFIX;
        }

        private static void ParsePacket(string packet)
        {
            Int32 packetIndex = 0;
            char opcode = '\0';
            byte hi_nible = 0;
            List<byte> dataBuffer = new List<byte>();
            foreach (char c in packet)
            {
                if (PREFIX == c) packetIndex = 0;
                else packetIndex++;

                if (packetIndex == 1)
                {
                    opcode = c;
                    dataBuffer.Clear();
                }

                if (packetIndex > 1)
                {
                    if (POSTFIX == c)
                    {
                        switch (opcode)
                        {
                            case OPCODE_RECEIVE:
                                string packetInfo = "Receive Packet\r\n";
                                packetInfo += $"Handle: {dataBuffer[0]}\r\nData: ";
                                for (int i = 1; i < dataBuffer.Count; i++) packetInfo += dataBuffer[i].ToString("X2");
                                packetInfo += "\r\n";
#if DEBUG
            System.Diagnostics.Debug.WriteLine(packetInfo);
#else
                                Console.WriteLine(packetInfo);
#endif
                                byte[] coreInterfacePacket = new byte[dataBuffer.Count - 1];
                                Array.Copy(dataBuffer.ToArray(), 1, coreInterfacePacket, 0, dataBuffer.Count - 1);
                                QSenseExamples.CoreInterfaceParser.ParsePacket();
                                break;
                            case OPCODE_STATUS:
                                string[] statusValues = new string[] { "Idle", "Scanning", "Connected" };
                                string statusInfo = "Status Packet\r\n";
                                statusInfo += $"Max. connections: {dataBuffer[0]}";
                                for (int i = 1; i < dataBuffer.Count; i++)
                                {
                                    statusInfo += $"Handle {i}: {statusValues[dataBuffer[i]]}";
                                }
#if DEBUG
            System.Diagnostics.Debug.WriteLine(statusInfo);
#else
                                Console.WriteLine(statusInfo);
#endif
                                break;
                            default:
                                break;
                        }

                        packetIndex = 0;
                    }
                    else if (ValidHex(c))
                    {
                        if (packetIndex % 2 == 0)
                        {
                            hi_nible = (byte)c;
                        }
                        else
                        {
                            dataBuffer.Add(GetValFromHexChars(hi_nible, (byte)c));
                        }
                    }
                    else
                    {
                        packetIndex--;
                    }
                }
            }
        }
        private static bool ValidHex(char c)
        {
            return (c >= '0' && c <= '9') || (c >= 'A' && c <= 'F') || (c >= 'a' && c <= 'f');
        }

        private static byte GetValFromHexChars(byte hi_nible, byte lo_nible)
        {
            Int32 result = 0;
            if (hi_nible >= '0' && hi_nible <= '9') result = hi_nible - '0';
            else if (hi_nible >= 'A' && hi_nible <= 'F') result = 10 + hi_nible - 'A';
            else if (hi_nible >= 'a' && hi_nible <= 'f') result = 10 + hi_nible - 'a';
            result = result << 4;
            if (lo_nible >= '0' && lo_nible <= '9') result += lo_nible - '0';
            else if (lo_nible >= 'A' && lo_nible <= 'F') result += 10 + lo_nible - 'A';
            else if (lo_nible >= 'a' && lo_nible <= 'f') result += 10 + lo_nible - 'a';
            return (byte)result;
        }
    }
}

File: Python Examples/DongleSerialInterfaceParser.py
import CoreInterfaceParser
class DongleSerialInterfaceParser:
    MAX_NODES = 13
    DEVICE_NAME = "QSense\0"
    PREFIX = "$"
    POSTFIX = "\n"

    OPCODE_TRANSMIT = "T"
    OPCODE_RECEIVE = "R"
    OPCODE_CONNECT = "C"
    OPCODE_CONNECT_WHITELIST = "W"
    OPCODE_DISCONNECT = "D"
    OPCODE_STOPSCAN = "I"
    OPCODE_STATUS = "S"

    @staticmethod
    def byte_array_to_string(data: bytes) -> str:
        return data.hex().upper()

    @staticmethod
    def create_transmit_packet(core_interface_packet: bytes, handle: int) -> str:
        hex_string = DongleSerialInterfaceParser.byte_array_to_string(core_interface_packet)
        return f"{DongleSerialInterfaceParser.PREFIX}{DongleSerialInterfaceParser.OPCODE_TRANSMIT}{handle:02X}{hex_string}{DongleSerialInterfaceParser.POSTFIX}"

    @staticmethod
    def create_connect_packet() -> str:
        packet = f"{DongleSerialInterfaceParser.PREFIX}{DongleSerialInterfaceParser.OPCODE_CONNECT}"
        packet += f"{DongleSerialInterfaceParser.MAX_NODES:02X}"
        buffer = DongleSerialInterfaceParser.DEVICE_NAME.encode("ascii")
        packet += buffer.hex().upper()
        return packet + DongleSerialInterfaceParser.POSTFIX

    @staticmethod
    def create_connect_whitelist_packet(serial_numbers: list[str]) -> str:
        packet = f"{DongleSerialInterfaceParser.PREFIX}{DongleSerialInterfaceParser.OPCODE_CONNECT_WHITELIST}"
        packet += f"{len(serial_numbers):02X}"
        for sn in serial_numbers:
            mac_address = int(sn.split("-")[0], 16)
            buffer = mac_address.to_bytes(8, "little")[:6] 
            packet += buffer.hex().upper()
        return packet + DongleSerialInterfaceParser.POSTFIX

    @staticmethod
    def create_disconnect_packet() -> str:
        return f"{DongleSerialInterfaceParser.PREFIX}{DongleSerialInterfaceParser.OPCODE_DISCONNECT}{DongleSerialInterfaceParser.POSTFIX}"

    @staticmethod
    def create_stop_scan_packet() -> str:
        return f"{DongleSerialInterfaceParser.PREFIX}{DongleSerialInterfaceParser.OPCODE_STOPSCAN}{DongleSerialInterfaceParser.POSTFIX}"

    @staticmethod
    def create_status_request() -> str:
        return f"{DongleSerialInterfaceParser.PREFIX}{DongleSerialInterfaceParser.OPCODE_STATUS}{DongleSerialInterfaceParser.POSTFIX}"

    @staticmethod
    def parse_packet(packet: str):
        packet_index = 0
        opcode = None
        hi_nibble = 0
        data_buffer: list[int] = []

        for c in packet:
            if c == DongleSerialInterfaceParser.PREFIX:
                packet_index = 0
            else:
                packet_index += 1

            if packet_index == 1:
                opcode = c
                data_buffer.clear()

            if packet_index > 1:
                if c == DongleSerialInterfaceParser.POSTFIX:
                    if opcode == DongleSerialInterfaceParser.OPCODE_RECEIVE:
                        packet_info = "Receive Packet\n"
                        packet_info += f"Handle: {data_buffer[0]}\nData: "
                        packet_info += "".join(f"{b:02X}" for b in data_buffer[1:])
                        print(packet_info)

                        core_interface_packet = bytes(data_buffer[1:])
                        CoreInterfaceParser.parse_packet(core_interface_packet)

                    elif opcode == DongleSerialInterfaceParser.OPCODE_STATUS:
                        status_values = ["Idle", "Scanning", "Connected"]
                        status_info = f"Status Packet\nMax. connections: {data_buffer[0]}"
                        for i in range(1, len(data_buffer)):
                            status_info += f"\nHandle {i}: {status_values[data_buffer[i]]}"
                        print(status_info)

                    packet_index = 0
                elif DongleSerialInterfaceParser.valid_hex(c):
                    if packet_index % 2 == 0:
                        hi_nibble = ord(c)
                    else:
                        data_buffer.append(
                            DongleSerialInterfaceParser.get_val_from_hex_chars(hi_nibble, ord(c))
                        )
                else:
                    packet_index -= 1

    @staticmethod
    def valid_hex(c: str) -> bool:
        return ('0' <= c <= '9') or ('A' <= c <= 'F') or ('a' <= c <= 'f')

    @staticmethod
    def get_val_from_hex_chars(hi_nibble: int, lo_nibble: int) -> int:
        def hex_val(ch: int) -> int:
            if ord("0") <= ch <= ord("9"):
                return ch - ord("0")
            elif ord("A") <= ch <= ord("F"):
                return 10 + ch - ord("A")
            elif ord("a") <= ch <= ord("f"):
                return 10 + ch - ord("a")
            return 0

        return (hex_val(hi_nibble) << 4) + hex_val(lo_nibble)
